### ISSUE TITLE
Distribution Strategy

### ISSUE THREAD

**Issue by IanMayo** [debrief/pepys-import#44]
Once deployed on the user network, ideally the an import will happen as follows:
* user opens command prompt
* user navigates to folder containing data
* user enters `pepys_import`

`Pepys_import` will be a `.bat` file on their path. It will instruct Python to run the `file_processor` with the current folder as an argument.

The %cd% bat-file variable contains the directory the current working directory: https://stackoverflow.com/a/4420078/92441

For evaluation purposes, the data will be imported into SQLite database named `pepys_import.db`, in the current working folder.

Our distribution strategy should also include:
* process to build a GitHub release (zip file) that contains all dependencies, sufficient that it can run on a PC without an Internet connection
* process to deploy release to shared folder
* script to make `pepys_import` available to user (including use of virtual environment)

**Update:** This morning I came across a serious improvement to the above.  The MS-Windows context menu has a "Send to" sub-menu.  It lists shared drives, but also apps such as "Dropbox". This would be an ideal integration for Pepys-import:
* user navigate to folder (or files) to import
* right-click on folder, select `Send to > Pepys Datastore`
* command prompt opens, user provides missing data, and/or allow/rejects import of files from that folder.

For this "Send to" support, the onsite IT team believe they can "push" the shortcut to all users.  They suggested that when we deliver our first release (_maybe_ a beta), they're happy to experiment with implementing this.

**Comment by robintw** [comment]
As part of the deployment, will users need to be able to import the various pepys libraries for use in their own scripts? Eg. once they've run `pepys-import` to import the data, will they need to use our code for accessing and manipulating the data?

If so, presumably they will be able to activate our virtual environment first, and then run their code that requires pepys libraries in our virtual environment?

**Comment by IanMayo** [comment]
Hi there - no, we've got an alternate strategy, though it's currently only in my imagination.

Pepys is deployed with a folder of core parsers - the ones that we maintain.  At runtime, Pepys loops through this folder, and registers the parsers.
It also checks for the presence of an environmental variable %PEPYS_USER_PARSERS%.  If this env variable is present, it loops through the contents of that folder, and registers the parsers from that folder.

So, each time Pepys runs, it loads the core and the user-controlled parsers.

Having the user parsers in the environmental variable means they can be maintained under version control in a tidy network location.

So, users will only be _using_ pepys in their own scripts as they write new parsers, but they're only interacting with a fairly narrow part of the API.

(I hope that all makes sense)

**Comment by robintw** [comment]
I've just realised that if they're going to be developing parsers using the pepys API then they'll need to develop and test those in a Python environment that has access to pepys-import. I've added a little bit to the bottom of the deployment plan to cover this (see https://docs.google.com/document/d/1h4LB_7Va-yOyg1vJywpZ1wKdt9-YeYX_R1FeFWu6ZBw/edit?ts=5e43dc93#)

**Comment by IanMayo** [comment]
I've just been considering the folder structure in your deployment plan.

If there are going to be any scripts in the top level, then they should be under version control. But, really it would be neater if they were in `pepys-import` repo.

Here's an alternate strategy:
```
- pepys_import
    - bin
        - python [*]
        pepys_import.bat
        pepys_admin.bat
        configure.bat
        configure.sh
    - libs
        - mod_spatialite [*]
        - sqlite [*]
    - pepys_import
    - tests
    - venv [*]
```
**Note:** folders with [*] are not under version control (_.git_ignore_)

The above strategy has the benefit that we can be very confident of the version of python, and we can even overwrite python DLLS if necessary.

Since we're using a 'private' instance of Python, I _don't think_ we need to worry about wheel or requirements.txt at deployment time.

`Configure` scripts can download/unpack the dependencies above (`curl/wget`) for developers.  We can make releases by just zipping up the above folder on an MS-Windows 10 PC/VM.

**Comment by robintw** [comment]
[pepys-import.zip](https://github.com/debrief/pepys-import/files/4232841/pepys-import.zip)

Attached is a first go at a deployment zip file. You should be able to unzip this file on any Windows 10 machine, and then run any of the batch files inside `pepys-import\bin` and they should work. This will be entirely independent of any other Python installation you may (or may not) have on that machine.

I have tested it on my bare Windows 10 VM, and on my main Windows machine at home - but there is always the possibility it won't work when tested - so have a go and let me know.

**Comment by IanMayo** [comment]
I'm seriously excited about trying this tomorrow!

**Comment by IanMayo** [comment]
Closing, Fixed in #89
